syntax = "proto3";

package parallax.confidence;

option go_package = "github.com/parallax/proto/confidence";

import "google/protobuf/timestamp.proto";
import "google/protobuf/struct.proto";
import "google/protobuf/empty.proto";

// Core confidence result message
message ConfidenceResult {
  string value_json = 1;                     // JSON-encoded result value
  double confidence = 2;                     // Confidence level (0.0-1.0)
  string agent_id = 3;                       // ID of the agent that produced this result
  google.protobuf.Timestamp timestamp = 4;   // When the result was produced
  repeated string uncertainties = 5;         // List of uncertainties/caveats
  string reasoning = 6;                      // Explanation of the result
  map<string, string> metadata = 7;          // Additional metadata
}

// Agent capabilities
message Capabilities {
  string agent_id = 1;
  string name = 2;
  repeated string capabilities = 3;          // List of capability tags
  double expertise_level = 4;                // Overall expertise (0.0-1.0)
  map<string, double> capability_scores = 5; // Per-capability confidence scores
}

// Health check response
message Health {
  enum Status {
    UNKNOWN = 0;
    HEALTHY = 1;
    UNHEALTHY = 2;
    DEGRADED = 3;
  }
  
  Status status = 1;
  string message = 2;
  google.protobuf.Timestamp last_check = 3;
  map<string, string> details = 4;
}

// Agent request for analysis
message AgentRequest {
  string task_id = 1;                        // Unique task identifier
  string task_description = 2;               // What to analyze
  google.protobuf.Struct data = 3;           // Input data as JSON
  map<string, string> context = 4;           // Additional context
  int32 timeout_ms = 5;                      // Timeout in milliseconds
  string pattern_name = 6;                   // Optional pattern to use
}

// Service definition for confidence-aware agents
service ConfidenceAgent {
  // Analyze a task and return result with confidence
  rpc Analyze(AgentRequest) returns (ConfidenceResult);
  
  // Get agent capabilities
  rpc GetCapabilities(google.protobuf.Empty) returns (Capabilities);
  
  // Health check
  rpc HealthCheck(google.protobuf.Empty) returns (Health);
  
  // Stream analysis for long-running tasks
  rpc StreamAnalyze(AgentRequest) returns (stream ConfidenceResult);
}